### PR TITLE
feat: Listen for CDC changed and update Master Doctor Index

### PR DESCRIPTION
**Pull request by CaiWillis**
TIS21-2748: Add endpoints/listeners Integration service to update master index

### CODE REVIEW

**Comment by CaiWillis** [comment]
integration_cdc_classes

File: src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/handler/CdcMessageHandler.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import com.mongodb.client.model.changestream.ChangeStreamDocument;
+import com.mongodb.client.model.changestream.OperationType;
+import org.springframework.stereotype.Component;
+import uk.nhs.hee.tis.revalidation.integration.cdc.service.CdcService;
+import uk.nhs.hee.tis.revalidation.integration.message.MessageHandler;
+
+import javax.naming.OperationNotSupportedException;
+
+@Component
+public class CdcMessageHandler<T extends ChangeStreamDocument> implements MessageHandler<T> {
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/handler/CdcMessageHandler.java:33]: Looking at the use of this I wasn't sure it should be a `@Component` and probably an `abstract class`.
:point_up: that seems consistent with your image describing it as a "Base Class".

File: src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/listener/CdcSqsMessageListener.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+public class CdcSqsMessageListener{
+  private CdcRecommendationMessageHandler cdcRecommendationMessageHandler;
+  private CdcDoctorMessageHandler cdcDoctorMessageHandler;
+
+  public CdcSqsMessageListener() {
+    this.cdcRecommendationMessageHandler = cdcRecommendationMessageHandler;
+    this.cdcDoctorMessageHandler = cdcDoctorMessageHandler;
+  }
+
+  @SqsListener(value = "${}")
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/listener/CdcSqsMessageListener.java:44]: I think `value = ` is redundant.  `@SqsListener("${}")` should do.
I'm not familiar with these annotations so don't know if there is a property name that you'll be adding.

File: src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/listener/CdcSqsMessageListener.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  public void getRecommendationMessage(ChangeStreamDocument<Recommendation> message) {
+    try{
+      cdcRecommendationMessageHandler.handleMessage(message);
+    } catch (Exception e) {
+      log.error(e.getMessage());
+    }
+  }
+
+  @SqsListener(value = "${}")
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/listener/CdcSqsMessageListener.java:53]: `@SqsListener("${}")` should do.

**Review by JosephKelly — COMMENTED** [review]
I would have liked to 'check my working' more but I think my comments are right.

I'd forget if it were me so I'm just acknowledging your previous message about copyrights.

File: src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/listener/CdcSqsMessageListener.java (hunk 2 above)
Comment (CaiWillis) [reply src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/listener/CdcSqsMessageListener.java:44]: Whoops! That's definitely supposed to have a value!

File: src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/handler/CdcMessageHandler.java (hunk 1 above)
Comment (CaiWillis) [reply src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/handler/CdcMessageHandler.java:33]: So this class is mostly to reduce code re-use - handle message is always going to have the same logic so don't think it needs to be abstract, and the subclasses are mostly used to inject the correct type and CdcService implementation - what do you think?

File: build.gradle [hunk 4]
```diff
@@ -84,6 +84,9 @@ dependencies {
 
   // Faker
   implementation "com.github.javafaker:javafaker:1.0.2"
+
+  // MongoDB
+  implementation 'org.mongodb:mongodb-driver-sync:4.5.1'
```
[the remark this answers is not in the document]
Comment (CaiWillis) [reply build.gradle:89]: Unfortunately this relies quite heavily on ChangeStreamDocument for which I think the library is mongodb-driver-sync

[the remark this answers is not in the document]
Comment (JosephKelly) [reply build.gradle:89]: ChangeStreamDocument looks like it's in `org.mongodb:mongodb-driver-core:4.5.1`.  Not sure what/if difference that would make

File: src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/listener/CdcSqsMessageListener.java (hunk 2 above)
Comment (DorisWongNHS) [reply src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/listener/CdcSqsMessageListener.java:52]: Just FYR, the queues for Preprod are created with Terrafrom, and here is the param name for the queue `/tis/revalidation/documentdb-cdc/preprod/queue-url/recommendation` (if no objection about the queue name). You may update application.yml and task-definition.json for this if you want.
Related PR: https://github.com/Health-Education-England/TIS-OPS/pull/434/files

File: src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/handler/CdcMessageHandler.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+import com.mongodb.client.model.changestream.ChangeStreamDocument;
+import com.mongodb.client.model.changestream.OperationType;
+import javax.naming.OperationNotSupportedException;
+import org.springframework.stereotype.Component;
+import uk.nhs.hee.tis.revalidation.integration.cdc.service.CdcService;
+import uk.nhs.hee.tis.revalidation.integration.message.MessageHandler;
+
+@Component
+public class CdcMessageHandler<T extends ChangeStreamDocument, U> implements MessageHandler<T> {
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/handler/CdcMessageHandler.java:32]: This is equivalent to `public class CdcMessageHandler<T> implements MessageHandler<ChangeStreamDocument<T>>`

You'd need to change the parameter type of handleMessage() from `T` to `ChangeStreamDocument<T>` but I thought I'd see if you think it's simpler / more readable that way.

**Review by JosephKelly — COMMENTED** [review]
It looks like whitespace formatting needs to be applied to the MessageHandler classes (& maybe elsewhere in changed files.

I unresolved 1 thread about the dependency but it's nearing that point where I feel like I should have invested more effort in making a recommendation over and above pointing at things that might need to be looked at.

File: src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/handler/CdcMessageHandler.java (hunk 1 above)
Comment (JosephKelly) [reply src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/handler/CdcMessageHandler.java]: I am struggling to see where it actually gets used.  The subclasses get pulled in but there doesn't seem to be a used of CdcMessageHandler that is used.

File: src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/handler/CdcMessageHandler.java (hunk 5 above)
Comment (CaiWillis) [reply src/main/java/uk/nhs/hee/tis/revalidation/integration/cdc/message/handler/CdcMessageHandler.java:32]: Thanks! that was bothering me

**Review by DorisWongNHS — APPROVED** [review]

**Review by JosephKelly — DISMISSED** [review]
Sorry I still haven't given as thorough a review of the tests as I would have liked but much happier with the code :-)

Haven't merged as I don't know if parameters are in place

File: .aws/task-definition-prod.json [hunk 6]
```diff
@@ -158,6 +158,14 @@
         {
           "name": "SQS_SYNC_GMC_QUEUE",
           "valueFrom": "/tis/revalidation/sync-gmc/prod/queue-url"
+        },
+        {
+          "name": "SQS_CDC_DOCTOR_QUEUE",
+          "valueFrom": "tis/revalidation/documentdb-cdc/prod/queue-url/doctorsfordb"
```
Comment (DorisWongNHS) [inline .aws/task-definition-prod.json:164]: Missing `/` at the front of the param name

File: .aws/task-definition.json [hunk 7]
```diff
@@ -158,6 +158,14 @@
         {
           "name": "SQS_SYNC_GMC_QUEUE",
           "valueFrom": "/tis/revalidation/sync-gmc/preprod/queue-url"
+        },
+        {
+          "name": "SQS_CDC_DOCTOR_QUEUE",
+          "valueFrom": "tis/revalidation/documentdb-cdc/preprod/queue-url/doctorsfordb"
```
Comment (DorisWongNHS) [inline .aws/task-definition.json:164]: Missing `/` at the front of the param name

**Review by DorisWongNHS — CHANGES_REQUESTED** [review]

**Comment by DorisWongNHS** [comment]
@JosephKelly good call. While I double check the param name, I found a type of missing `/`. I had a check from the param store, both queues for Preprod are in place on AWS, and I am applied the queues to Prod by Terraform with this PR: https://github.com/Health-Education-England/TIS-OPS/pull/438

**Review by JosephKelly — APPROVED** [review]

**Review by DorisWongNHS — APPROVED** [review]

**Review by jayanta2018 — APPROVED** [review]

**Pull request merged**